# Modellierungsprojekt Optimierungsverfahren, Modellierung und Simulation

## Organisatorisches
In dieser Übung sollen Sie verschiedene dynamische Energiespeicherprobleme als gemischt-ganzzahlige Programme modellieren, mit Gurobi lösen und die Lösung visualisieren und interpretieren.

Die Abgabe erfolgt in Form eines Jupyter-Notebooks. Nutzen Sie den Markup-Modus, um ihre mathematische Modellierung zu beschreiben. Mathematische Formeln können Sie zwischen zwei `$`-Zeichen oder zwei doppelten Dollarzeichen `$$` setzen. Sie können auch per Copy-and-paste Markup-Codeschnipsel kopieren, z.B. aus den Übungsblättern.

Die Abgabe erfolgt in Teams von bis zu 4 Personen, die in ILIAS registriert werden müssen. Die Notebooks müssen bis **Montag, 19.06.2022, 18:00** in ILIAS abgegeben werden. 

Die Veranstaltungstermine am 09.06. sowie in der Woche ab dem 12.06. sind für die Arbeit am Projekt vorgesehen. Ich bin während der Termine vor Ort und stehe für Rückfragen und Unterstützung zur Verfügung. In den Veranstaltungsterminen in der Woche ab dem 19.06.2023 werde ich Sie in Einzelgesprächen (in Präsenz) zu ihrer abgegebenen Lösung befragen (ca. 10-15 Minuten). Die Termine bekommen Sie separat mitgeteilt.

Insgesamt gibt es 100 Punkte auf das Projekt, wobei 50 Punkte für eine 4.0 erforderlich sind. Folgende Aspekte werden für die Punktevergabe berücksichtigt:
- Korrektheit des mathematischen Modells
- Korrektheit der Implementierung
- Interpretation der Ergebnisse
- Darstellung, Übersichtlichkeit und Verständlichkeit des abgegebenen Jupyter Notebooks
- Einzelgespräche

Die Punktzahlen der einzelnen Aufgaben (z.B. 10+5) beziehen sich auf Punkte für die Abgabe plus Punkte für die (mündliche) Beantwortung von Fragen zu den Aufgaben in den Einzelgesprächen. Mögliche Fragen sind z.B.:

- "Wie haben Sie Bedingung X aus der Aufgabe im Modell realisiert?"
- "Was bedeutet die Constraint Y in Ihrem Modell?"
- "Was bedeuten die Terme in der Zielfunktion?"
- etc.

Das Projekt macht 30\% der Modulnote aus.

## Ausgangssituation
Sie sind Assistent der Geschäftsführung in einem mittelständischen Unternehmen. An einem Standort Ihrer Firma wurde kürzlich eine Photovoltaikanlage auf den Dächern der Gebäude sowie ein 100 kWh Stromspeicher (Batterie) installiert. Die PV-Anlage soll vorrangig dafür verwendet werden, um den Strombedarf des Standorts zu decken. Wenn die PV-Anlage mehr Strom produziert als der Standort gerade benötigt, wird damit der Stromspeicher geladen oder der Strom ins Netz eingespeist. Für das Einspeisen bekommen Sie eine Vergütung von 4 ct/kWh. Falls der Standort mehr Strom benötigt als die PV-Anlage produziert und die Batterie leer ist, so beziehen Sie Strom für 22 ct/kWh aus dem Netz. Schematisch sieht das Ganze so aus:

<div>
<img src="Aufgabe1.png" width="750"/>
</div>

Die Batterie hat eine Effizienz von 90% beim Be- und Entladen. Das bedeutet, dass für jede kWh Strom, die in der Batterie gespeichert wird, der Ladezustand um 0.9 kWh steigt. Umgekehrt kann mit einer 1 kWh aus der Batterie nur 0.9 kWh Bedarf gedeckt werden. Die Batterie kann mit einer Maximalleistung von 50kW ge- bzw. entladen werden, d.h. während einer Stunde muss die Summe aller Lastflüsse in und aus der Batterie kleiner als 50 kWh sein.

Ihr Chef hat gelesen, dass es mit Hilfe eines dynamischen Strompreistarifs möglich ist, den Betrieb der Anlage und die Energieversorgung des Standorts zu optimieren. Er bittet Sie, der Sache nachzugehen.

## ✏ Aufgabe 0: Datenanalyse (5 Punkte)
Sie möchten zunächst das System abbilden, wie es heute arbeitet. Dafür besorgen Sie sich historische Daten über den Stromverbrauch des Standorts (`Last [kWh]`) sowie Daten über die Sonneneinstrahlung, aus der Sie die Stromproduktion der PV-Anlage errechnen (`Solar [kWh]`). Alle Daten sind stundengenau für ein Jahr in der Datei `energy_2020.pkl` hinterlegt. In der Spalte `Preis [EUR/kWh]` ist zusätzlich der variable Strompreis für jede Stunde gegeben. Diesen benötigen Sie erst später.

In [1]:
import pandas as pd
import numpy as np  
import plotly.express as px
df = pd.read_pickle("energy_2020_corrected.pkl")
df

Datum (MEZ)  Solar [kWh]  Preis [EUR/kWh]  Last [kWh]
0    2020-01-01 00:00:00+01:00          0.0          0.20940     94.2636
1    2020-01-01 01:00:00+01:00          0.0          0.19300     92.4732
2    2020-01-01 02:00:00+01:00          0.0          0.18275     90.2108
3    2020-01-01 03:00:00+01:00          0.0          0.16160     87.5294
4    2020-01-01 04:00:00+01:00          0.0          0.15425     86.8490
...                        ...          ...              ...         ...
8779 2020-12-31 19:00:00+01:00          0.0          0.29735    120.8536
8780 2020-12-31 20:00:00+01:00          0.0          0.28350    112.4558
8781 2020-12-31 21:00:00+01:00          0.0          0.26220    106.1954
8782 2020-12-31 22:00:00+01:00          0.0          0.25930    103.6396
8783 2020-12-31 23:00:00+01:00          0.0          0.26130     98.7446

[8784 rows x 4 columns]

Bearbeiten Sie folgende Aufgaben:
1. Visualiseren Sie die drei Zeitreihen.
1. Wie hoch war der Gesamtstromverbrauch des Standorts im Jahr 2020?
1. Wie hoch wären die Gesamtstromkosten des Standorts im Jahr 2020 ohne PV-Anlage gewesen (d.h. mit 22 ct/kWh Strompreis)?
1. Wie hoch wäre die gesamte produzierte Strommenge der PV-Anlage im Jahr 2020 gewesen?
1. Wie hoch war der durchschnittliche Strompreis im Jahr 2020?

In [2]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# Erstellen der Subplots
fig = make_subplots(rows=3, cols=1)

# Hinzufügen der Line Plots zu den Subplots
fig.add_trace(go.Scatter(x=df["Datum (MEZ)"], y=df["Solar [kWh]"], mode='lines'), row=1, col=1)
fig.add_trace(go.Scatter(x=df["Datum (MEZ)"], y=df["Preis [EUR/kWh]"], mode='lines'), row=2, col=1)
fig.add_trace(go.Scatter(x=df["Datum (MEZ)"], y=df["Last [kWh]"], mode='lines'), row=3, col=1)

# Anpassung der Layout-Eigenschaften
fig.update_layout(height=800,  title_text="Visualisierung der Daten")

fig.update_traces(
    name="Solar [kWh]",
    row=1, col=1,
    hovertemplate="<b>Petal Length</b>: %{x}<br><b>Petal Width</b>: %{y}<extra></extra>"
)
fig.update_traces(
    name="Preis [EUR/kWh]",
    row=2, col=1,
    #hovertemplate="<b>Sepal Length</b>: %{x}<br><b>Sepal Width</b>: %{y}<extra></extra>"
)
fig.update_traces(
    name="Last [kWh]",
    row=3, col=1
)

# Anzeigen der Subplots
fig.show()

In [4]:
df.dtypes

Datum (MEZ)        datetime64[ns, pytz.FixedOffset(60)]
Solar [kWh]                                     float64
Preis [EUR/kWh]                                 float64
Last [kWh]                                      float64
dtype: object

In [8]:
df["Datum (MEZ)"].dt.year == 2020

0       True
1       True
2       True
3       True
4       True
        ... 
8779    True
8780    True
8781    True
8782    True
8783    True
Name: Datum (MEZ), Length: 8784, dtype: bool

In [11]:
#2.
df.loc[df["Datum (MEZ)"].dt.year == 2020]["Last [kWh]"].sum()

1043805.6564

In [12]:
#3.
df.loc[df["Datum (MEZ)"].dt.year == 2020]["Last [kWh]"].sum() * 0.22

229637.244408

In [14]:
#4.
df["Solar [kWh]"].sum()

970692.484

In [15]:
#5.
df["Preis [EUR/kWh]"].mean()

0.1523535746812386

## ✏ Aufabe 1: Basismodell (25+10 Punkte)
Modellieren Sie das System als lineares Programm, so dass die Gesamtstromkosten minimiert werden. Der Strombedarf des Standorts muss zu jedem Zeitpunkt 100% gedeckt werden. Gehen Sie vereinfachend davon aus, dass die Batterie zu Beginn des betrachteten Zeitraums eine Ladestand von 0 hat. 

1. Formulieren Sie ein lineares Programm in mathematischer Notation an, welches das Problem löst (Indexmengen, Problemdaten, Variablen, Zielfunktion, Nebenbedingungen).
2. Lösen Sie das Problem mit Gurobi, lesen Sie folgende Größen aus der Lösung aus, und visualisieren Sie sie:
   - Ladezustand der Batterie zu jedem Zeitpunkt $t$.
   - Menge an ein- bzw. ausgespeichertem Strom zu jedem Zeitpunkt $t$.
   - Strombezug und Stromverkauf an den Netzbetreiber zu jedem Zeitpunkt $t$.
   - Menge an selbst verbrauchtem Strom aus der PV-Anlage zu jedem Zeitpunkt $t$.
3. Geben Sie außerdem für das Jahr 2020 die folgenden Kenngrößen an:
   - Eigenverbrauch gesamt (in kWh)
   - Eigenverbrauch aus der Batterie (in kWh)
   - Gesamtstromkosten des Standorts mit PV-Anlage und Batterie (in EUR)
   - Ersparnis durch Eigenverbrauch gesamt (in EUR)
   - Ersparnis durch Eigenverbrauch aus der Batterie (in EUR)

$$
x_{ij}=\begin{cases}
1 & \text{Arbeitsaufgabe $i$ wird auf Maschine $j$ ausgeführt} \\
0 & \, \text{sonst}
\end{cases}
$$

Vollständiges MILP:
\begin{alignat*}{5}
\min          & \quad  &   k_{8784} &          \\[4mm]
\text{s.t. } & &  \sum_{i=1}^n d_ix_{ij} & \leq y,\quad\text{für }j=1,\dots, 4  \\[2mm]
& &  \sum_{i=1}^4 x_{ij} & = 1,\quad\text{für }i=1,\dots, n \\[2mm]
& &  \sum_{i=1}^4 x_{ij} & = 1,\quad\text{für }i=1,\dots, n \\[2mm]
& &  x_{ij} & \in \{0,1\},\quad\text{für }i=1,\dots, n,\ j=1,\dots,4
\end{alignat*}

\begin{alignat*}{5}
\min          & \quad  &   k_{8784} &          \\[4mm]
% Weitere Gleichungen und Restriktionen können hier eingefügt werden
k_{t+1} &= k_t + 0.22 \cdot E_t - 0.04 \cdot s_{vt} \\
% Weitere Gleichungen und Restriktionen können hier eingefügt werden
\end{alignat*}

# Aufgabe1
Formulieren Sie die Problemstellung als ganzzahliges lineares Programm.

Entscheidungsvariablen:

$l_t$ Steht für unseren Strombedarf den unsere Anlage zu dem Zeitpunkt t benötigt.

$s_t$ Steht für unseren erzeugte Strommenge durch unsere Solaranlage zu dem Zeitpunkt t.

$s_t^v$ Steht für unsere Strommmenge, die wir zu dem Zeitpunkt t verkaufen.

$s_t^e$ Steht für unsere Strommmenge an Solarstrom, die wir zu dem Zeitpunkt t verbrauchen.

$p_t^+$ Steht für den Preis den wir für unseren Strom zu dem Zeitpunkt t bekommen.

$p_t^-$ Steht für den Preis den wir für unseren Strom zu dem Zeitpunkt t bezahlen.

$k$ Steht für unseren Stromkosten.

$E_t$ Steht für unsere eingekaufte Strommenge zu dem Zeitpunkt t.

$b$ Steht für den Ladestand unserer Batterie.

$b_t^+$ Steht für den die Strommenge die in unsere Batterie fließt.

$b_t^-$ Steht für den die Strommenge die aus unserer Batterie fließt.



Vollständiges MILP:
\begin{alignat*}{5}
\min          & \quad  &   k_{8784} &          \\[4mm]
% Weitere Gleichungen und Restriktionen können hier eingefügt werden
k_{t+1} &= k_t + 0.22 \cdot E_t - 0.04 \cdot s_t^v \\
E_t &= l_t - 0.9 \cdot b_t^- - s_t^e \\
E_t &\geq 0 \\
b_{t+1} &= b_t + 0.9 \cdot b_t^+ - b_t^- \\
b_{t+1} &\leq 100 \\
b_{t+1} &\geq 0 \\
s_t &= s_t^v + s_t^e + b_t^+ \\
{b_t^+ \over 0.9}   + b_t^- &\leq 50 \\
b_t^- &\geq 0 \\
b_t^+ &\geq 0 \\
b_0 &= 0 \\
% Weitere Gleichungen und Restriktionen können hier eingefügt werden
\end{alignat*}


In [ ]:
# TO DO

## ✏ Aufgabe 2: Variabler Strompreis (10+5 Punkte)
Sie haben nun eine realistische Baseline, gegen die Sie alle weiteren Optionen vergleichen können. Als nächstes möchten Sie das Potential des dynamischen Strompreistarifes berechnen. Sie haben sich dafür bereits die relevanten historischen Strompreisdaten besorgt (Spalte `Preis [EUR/kWh]`). Der Tarif wirkt zunächst nur in eine Richtung, d.h. der Einkaufspreis ist variabel, aber die Einspeisevergütung liegt nach wie vor bei konstant 4 ct/kWh. Es kann sich durchaus lohnen, Energie aus dem Netz zu beziehen und in die Batterie einzuspeichern, wenn der Strompreis gerade niedrig ist. 

Insgesamt ergibt sich folgendes Bild:

<div>
<img src="Aufgabe2.png" width="750"/>
</div>

Erweitern Sie nun ihr Modell aus Aufgabe 1 und entwickeln Sie ein lineares Programm (mathematische Formulierung, Formulierung als Python Code und Lösung mit Gurobi) für das optimale Ein- und Ausspeichern von Strom in den Batteriespeicher bei variablem Strompreis beim Stromkauf. Berechnen Sie dieselben Kenngrößen wie in Aufgabe 1 und vergleichen Sie die Lösungen. Heben Sie exemplarisch 2-3 typische Situationen (einige Stunden oder Tage) hervor, bei denen sich die Lösungen sichtbar unterscheiden.

In [ ]:
# TO DO

## ✏ Aufgabe 3: Dimensionierung der Batterie (10+5 Punkte)
Simulieren Sie das Einsparpotential für unterschiedliche Größen von Batteriespeichern. Variieren Sie dafür die Batteriegröße in 100 kWh-Schritten zwischen 0 und 1000 kWh und berechnen Sie für jede Größe das Einsparpotential (für jede Batteriegröße ist ein LP zu lösen).

In [ ]:
# TO DO

## ✏ Aufgabe 4: Einspeisung ins Netz (10+5 Punkte)
Sie haben nun ein gutes Verständnis des Einsparpotentials ihrer aktuellen Anlage gewonnen sowie der Potentiale, die sich durch unterschiedliche Batteriespeicher realisieren lassen könnten. Bisher gehen Ihre Modelle von einer konstanten Einspeisevergütung aus. Untersuchen Sie nun auch noch das Potential, das sich bietet, wenn Sie zusätzlich als *Stromlieferant* am Strommarkt teilnehmen:

<div>
<img src="Aufgabe4.png" width="750"/>
</div>

Wir nehmen an, dass dies unter folgenden Bedingungen möglich ist:
- Pro Zeiteinheit müssen Sie entscheiden, ob Sie Strom zum pauschalen Preis einspeisen möchten oder zum variablen Preis.
- Die Einspeisung kann aus der PV-Anlage und/oder der Batterie erfolgen.
- Falls Sie während einer Zeiteinheit zum variablen Preis ins Netz einspeisen möchten, so ist dies nur möglich, wenn Sie mindestens 100 kWh einspeisen (kombiniert aus PV-Anlage und Batterie).
- Falls Sie zum variablen Preis ins Netz einspeisen möchten, so ist dies zum Marktpreis abzüglich 5 ct/kWh möglich.

Erweitern Sie nun ihr Modell aus Aufgabe 2 und entwickeln Sie ein lineares Programm für das optimale Ein- und Ausspeichern von Strom in den Batteriespeicher bei variablem Strompreis beim Stromkauf und -verkauf (mathematische Formulierung, Formulierung als Python Code und Lösung mit Gurobi). Berechnen Sie dieselben Kenngrößen wie in Aufgabe 2 und vergleichen Sie die Lösungen. Heben Sie exemplarisch 2-3 typische Situationen (einige Stunden oder Tage) hervor, bei denen sich die Lösungen sichtbar unterscheiden.

In [ ]:
# TO DO

## ✏ Aufgabe 5: Validierung (10+5 Punkte)
Bei all Ihren Betrachtungen haben Sie bisher einen Aspekt außer Acht gelassen: Die Berechnungen wurden *a posteriori* durchgeführt, d.h. sie wurden im Nachhinein auf bekannten Daten angestellt. In der Realität weiß man aber zum Zeitpunkt der Entscheidung (ein-/ausspeichern) noch nicht, wie sich der Strompreis, der Verbrauch und die produzierte Strommenge zukünftig entwickeln werden. Das so berechnete Einsparpotential ist deshalb wahrscheinlich ein wenig zu optimistisch.

In der Datei `energy_2020_scenarios.pkl` finden Sie 100 verschiedene Szenarien, wie der Strompreis tatsächlich sein könnte. Evaluieren Sie anhand der verschiedenen Szenarien, wie hoch die Stromkosten mit dem optimalen Ein- und Ausspeichern wie in den Aufgaben 1, 2 und 4 berechnet tatsächlich sein könnten:
- Berechnen Sie für jedes Szenario und für jedes der drei Modelle die Stromkosten.
- Visualisieren Sie die Ergebnisse der Simulation in einem geeigneten (vergleichenden) statistischen Plot.

Was müssten Sie zusätzlich beachten, wenn Sie eine derartige Simulation für verschiedene Szenarien des Strombedarfs und der Stromerzeugung machen würden?

In [ ]:
# TO DO